src/reflect/scala/reflect/internal/Symbols.scala

```diff
@@ -1562,7 +1562,7 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
       */
     def substInfo(syms0: List[Symbol], syms1: List[Symbol]): this.type =
       if (syms0.isEmpty) this
-      else modifyInfo(_.substSym(syms0, syms1))
+      else modifyInfo(_.substSym(syms0, syms1, Some(this)))
```

Comment:
I don't understand this; can you explain? We avoid changing any symbol in the info to `this`?

Reply:
Yes, we avoid.  If anything in the info is turned to `this`, for cases where `this` is precisely part of the `syms1` list, then is creates a bad cycle.  My fix in `TypeMap` seemed to sometimes work too well, and a few partests were finding themselves with those reference cycles.

Reply:
I feel like this is fighting a symptom, not the root cause of the problem. I will take a look.

Reply:
One of the failing tests is https://github.com/scala/scala/blob/2.13.x/test/files/pos/t1210a.scala. I looked at this one in the debugger.

We have a `TypeRef` `tp = Test.this.T`. When calling `tp.mapOver(asSeenFrom)` with `asSeenFrom = AsSeenFromMap(OO.Test.<refinement>.type, class Test)`, we get `Test.this.v.T`. This is incorrect, as far as I can tell. The result should still be `Test.this.T`.

I'm sorry this is a bit of a can of worms :( I'll see if I can fix the above.

Reply:
Scratch that, I think the mapOver is correct. So for

```
abstract class Test { self =>
  type T
  val v: Test {type T' = self.T} = self.v.v
}
```

In the selection `self.v.v` we compute the type of `v` asSeenFrom (`Test.v.type`, `Test`)
  - we clone the decls in the refinement and get `Test { type T'' = Test.this.v.T' }`, which looks good
  - but then the `mapOver(syms: List[Symbol])` (which created new symbols for the decls in the refinement) calls substSym with `from=List(T')` and `to=List(T''). This changes the type to `Test { type T'' = T'' }` basically

Your solution might be the right one, I'm not sure.. @adriaanm any idea at which level we should break the cycle?

Reply:
Yeah so changing the order of 2. and 3. fixes the failing existing partests, but it also breaks the fix for scala/bug#10911, which is the point of this PR.. the reason is well explained in https://github.com/scala/bug/issues/10911#issuecomment-392589640.

Reply:
Maybe we should just go back to the previous solution that used your `cloneSymbolsWithModify`, i.e., revert parts of https://github.com/scala/scala/commit/aa7578210b74c3b051dbf49a9c3b5c2298d50f56, does that sound plausible @NirvanaNrv?

Reply:
From the start I had to include this piece preventing symbol cycles; it was with the cloneSymbolsWithModify change.  Having such symbol referenced in its type was leading to stack overflows in so many places that I just assumed it was plain forbidden by design, just as in the language itself.

The call to substSym happening after the cloning, I read it as part of that cloning.  And my point was to have the cloning happening after the TypeMap application.